lib/rex/proto/ldap/server.rb

```diff
@@ -0,0 +1,292 @@
+# -*- coding: binary -*-
+
+require 'rex/socket'
+require 'net/ldap'
+
+module Rex
+module Proto
+module LDAP
+
+class Server
+
+  module LdapClient
+    attr_accessor :authenticated
+    #
+    # Initialize LDAP client state
+    #
+    def init_ldap_client
+      self.authenticated = false
+    end
+  end
+
+  class MockLdapClient
+    attr_reader :peerhost, :peerport, :srvsock
+    #
+    # Create mock LDAP client
+    #
+    # @param host [String] PeerHost IP address
+    # @param port [Fixnum] PeerPort integer
+    # @param sock [Socket] Connection socket
+    def initialize(host, port, sock)
+      @peerhost = host
+      @peerport = port
+      @srvsock  = sock
+    end
+
+    #
+    # Test method to prevent GC/ObjectSpace abuse via class lookups
+    #
+    def mock_ldap_client?
+      true
+    end
+
+    def write(data)
+      srvsock.sendto(data, peerhost, peerport)
+    end
+  end
+
+  include Rex::IO::GramServer
+  #
+  # Create LDAP Server
+  #
+  # @param lhost [String] Listener address
+  # @param lport [Fixnum] Listener port
+  # @param udp [TrueClass, FalseClass] Listen on UDP socket
+  # @param tcp [TrueClass, FalseClass] Listen on TCP socket
+  # @param ldif [String] LDIF data
+  # @param ctx [Hash] Framework context for sockets
+  # @param dblock [Proc] Handler for :dispatch_request flow control interception
+  # @param sblock [Proc] Handler for :send_response flow control interception
+  #
+  # @return [Rex::Proto::LDAP::Server] LDAP Server object
+  attr_reader :serve_udp, :serve_tcp, :sock_options, :udp_sock, :tcp_sock, :syntax, :ldif
+  def initialize(lhost = '0.0.0.0', lport = 389, udp = true, tcp = true, ldif = nil, comm = nil, ctx = {}, dblock = nil, sblock = nil)
```

Comment:
Chances are we're going to want to have a SSL option for this pretty soon. There's a lot of positional arguments already and when we add SSL we'll need to add even more. It might scale better if we used keyword arguments here for maybe everything to the right of `lport`.

```suggestion
  def initialize(lhost = '0.0.0.0', lport = 389, udp: true, tcp: true, ldif: nil, comm: nil, context: {}, dblock: nil, sblock: nil)
```

Reply:
pre-negotiated TLS or some STARTTLS nonsense?

Reply:
Marking this as resolved because the kwargs I proposed broke compatibility with the service manager.